test/backends.py

```diff
@@ -0,0 +1,217 @@
+                if (isinstance(argument, self.usr_np.ndarray)
+                        and argument.shape == ()):
+                    return self.usr_np.abs(argument)
+                return self.usr_np.linalg.norm(argument, normord)
+
+        return JaxPyroWrapper
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.python.gen_thermochem_code(sol=sol)
+
+    @staticmethod
+    def extract_interface(module, mechname: str, user_np):
+        base_cls = module.Thermochemistry
+
+        if user_np == jnp:
+            base_cls = PythonBackend._gen_jax_wrapper(base_cls)
+
+        return base_cls(user_np)
+
+
+class CppBackend(Backend):
+    def get_name():
+        return "cpp"
+
+    def fix_arg(self, x):
+        if isinstance(x, Iterable) and not isinstance(x, str):
+            return self.pyro.VectorDouble(x)
+        return super().fix_arg(x)
+
+    def fix_result(self, result):
+        if isinstance(result, self.pyro.VectorDouble):
+            return np.array(result)
+        return super().fix_result(result)
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.cpp.gen_thermochem_code(sol=sol, namespace=name)
+
+
+class FortranBackend(Backend):
+    def get_name():
+        return "fortran90"
+
+    @staticmethod
+    def extract_interface(module, mechname: str, user_np):
+        return getattr(module, f"libpyro_fortran90_{mechname}")
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.fortran90.gen_thermochem_code(
+            sol=sol, module_name=name
+        )
+
+
+BACKENDS: typing.Dict[str, Backend] = {
+    CppBackend.get_name(): CppBackend,
+    FortranBackend.get_name(): FortranBackend,
+    PythonBackend.get_name(): PythonBackend
+}
+
+
+class Pyrometheus:
+    def __init__(self, mechname: str, backend: Backend, user_np):
+        self.backend = backend
+        self.pyro = self.backend.extract_interface(
+            importlib.import_module(f"libpyro_{self.backend.get_name()}_{mechname}"),
+            mechname,
+            user_np
+        )
+
+    def __getattr__(self, name: str):
+        attrib = getattr(self.pyro, name)
+        if not callable(attrib):
+            return attrib
+
+        def wrapper(*args, **kwargs):
+            assert (len(kwargs) == 0)
+
+            backend_instance = self.backend(self.pyro)
+
+            return backend_instance.fix_result(
+                attrib(*[backend_instance.fix_arg(x) for x in args])
+            )
+
+        return wrapper
+
+
+def pyro_init(mechname: str, user_np, request) -> (ct.Solution, Pyrometheus):
+    mech_dir = pathlib.Path(__file__).parent / "mechs"
+    sol = ct.Solution(mech_dir / f"{mechname}.yaml", "gas")
+    backend = BACKENDS[request.config.getoption("backend").lower()]
+
+    return sol, Pyrometheus(mechname, backend, user_np)
```

Comment:
Why is a `ct.Solution` a product of pyro initialization?

Reply:
It could but then you would need to prepend all `sol` uses with `ptk.` which I don't think is an improvement. I don't see how this is a _sin_... We would need to construct the solution anyway, so why not return it as well?

Reply:
Yes, that is cumbersome indeed.